lib/core/src/restore.rs

```diff
@@ -0,0 +1,820 @@
+        let mut btc_user_lockup_tx_ids: HashMap<String, Txid> = HashMap::new();
+        let mut btc_refund_tx_ids: HashMap<String, Txid> = HashMap::new();
+
+        for (swap_id, history) in chain_receive_histories_by_swap_id {
+            info!("[Recover Chain Receive] Checking swap {swap_id}");
+
+            match history.lbtc_claim_script_history.len() {
+                2 => {
+                    let first = &history.lbtc_claim_script_history[0];
+                    let second = &history.lbtc_claim_script_history[1];
+
+                    // If a history tx is a known incoming txs, it's the claim tx
+                    let (lockup_tx_id, claim_tx_id) =
+                        match tx_map.incoming_tx_map.contains_key::<Txid>(first) {
+                            true => (second, first),
+                            false => (first, second),
+                        };
+                    lbtc_server_lockup_claim_tx_ids
+                        .insert(swap_id.clone(), (*lockup_tx_id, *claim_tx_id));
+                }
+                n => {
+                    error!("L-BTC script history with unexpected length {n} found while recovering data for Chain Receive Swap {swap_id}")
+                }
+            }
+
+            match history.btc_lockup_script_history.len() {
+                // TODO How to treat case when history length > 2? (address re-use)
+                x if x >= 2 => {
+                    // TODO How to tell the user lockup tx apart from the refund tx? Is the order in which they're received from Electrum reliable?
+                    // TODO How to tell BTC refund apart from BTC server claim tx?
+                    btc_user_lockup_tx_ids
+                        .insert(swap_id.clone(), history.btc_lockup_script_history[0]);
+                    btc_refund_tx_ids.insert(swap_id.clone(), history.btc_lockup_script_history[1]);
+                }
+                n => {
+                    error!("BTC script history with unexpected length {n} found while recovering data for Chain Receive Swap {swap_id}")
+                }
+            }
+        }
+
+        info!("[Recover Chain Receive] Found {} L-BTC server lockup and claim tx pairs from onchain data", lbtc_server_lockup_claim_tx_ids.len());
+        info!(
+            "[Recover Chain Receive] Found {} BTC user lockup tx IDs from onchain data",
+            btc_user_lockup_tx_ids.len()
+        );
+        info!(
+            "[Recover Chain Receive] Found {} BTC refund tx IDs from onchain data",
+            btc_refund_tx_ids.len()
+        );
+
+        Ok(RecoveredOnchainDataChainReceive {
+            lbtc_server_lockup_claim_tx_ids,
+            btc_user_lockup_tx_ids,
+            btc_refund_tx_ids,
+        })
+    }
+}
+
+/// Methods to simulate the immutable DB data available from real-time sync
+// TODO Remove once real-time sync is integrated
+pub(crate) mod immutable {
```

Comment:
NIT: I don't know if we need another module here. Also immutable is too generic in this context IMO.

Reply:
This module (marked as `TODO Remove once real-time sync is integrated`) is a placeholder for the immutable DB interface.

It contains ways to "simulate" the immutable data (now deriving it from actual DB data) until the interface is available. I placed these helper methods in this module, so it's easy to remove them when the interface is ready.